randomname/util.py

```diff
@@ -188,21 +188,23 @@ def doalias(fname):
 
 def safepath(f):
     '''Make sure a path doesn't go up any directories.'''
-    return os.path.abspath('/' + f).lstrip('/')
+    name_parts = f.split('/')
+    name_parts = [part for part in name_parts if not part.startswith('..')]
+    return '/'.join(name_parts).lstrip('/')
```

Comment:
Actually this no longer makes the assumption as we've changed `getallcategories` according to your suggestion to format the paths via `/` (independent of the OS). I'm also pretty sure this is true as the CI passes.

Reply:
to illustrate
```python
WORD_PATH = 'C:\\...\..\wordlists'
name = 'noun/food'

# current

path = safepath(name + ".txt") # noun/food.txt
path = f'{WORD_PATH.rstrip("/")}/{path}'  # C:\\...\..\wordlists/noun/food.txt

# I don't understand why this would return True unless windows is fine with /
if not os.path.isfile(path):  
    ...

# what I'm proposing

path = os.path.abspath(os.sep + name.replace('/', os.sep)).lstrip(os.sep)  # noun\food
path = os.path.join(WORD_PATH, path + '.txt')  # C:\\...\..\wordlists\noun\food.txt
if not os.path.isfile(path):  # everything is backslashes
    ...
```

Reply:
Actually I just took a quick look at the python os docs and it looks like `abspath` will convert slashes to backslashes for you since it calls `normpath`

>normpath (which is used by abspath): On Windows, it converts forward slashes to backward slashes.

So you can probably just simplify to:
```python
path = os.path.abspath(os.sep + name).lstrip(os.sep)
```

Reply:
>Python will automatically smooth things out ... apparently this works due to os.altsep

But in the article you linked:

> Technically this code will still work on Windows because Python has a hack (link to altsep) where it will recognize either kind of slash when you call open() on Windows. But even still, you shouldn’t depend on that. Not all Python libraries will work if you use wrong kind of slash on the wrong operating system — especially if they interface with external programs or libraries.

So it sounds like it's safer to not rely on this hack? idk

>The thing with your approach is that we call safepath both with filepaths (having os.path) and user input (using /).

I haven't tested this, but according to [this](https://docs.python.org/3/library/os.path.html#os.path.normpath), `os.path.abspath('noun/food')` and `os.path.abspath(r'noun\food')` should both return `'noun\food'` on windows so this should be able to handle both cases, (even though the input should always be `'/'` from the user).

And anyways, `safepath` is only called by `as_valid_path` and `as_valid_path` is only called in two places:
```python
# load_file()
with open(as_valid_path(name, required=True), 'r') as f: ...
# here we want backslashes (although with os.altsep is doesn't technically matter)

# and close_matches()
[f for f in ('{}/{}'.format(part0, m) for m in _ms) if as_valid_path(f)]
# here we only care if it returns None or not
```
and neither place asks for `noun/food` to use forward slashes so I don't think there should be a problem.